src/diagnostics/MOM_spatial_means.F90

```diff
@@ -330,56 +363,65 @@ function global_mass_integral(h, G, GV, var, on_PE_only, scale, tmp_scale, unsca
+                                !! or [R Z L2 B ~> kg b] if both unscale and tmp_scale are present
 
   ! Local variables
-  ! In the following comments, [A] is used to indicate the arbitrary, possibly rescaled units of the
+  ! In the following comments, [A ~> a] is used to indicate the arbitrary, possibly rescaled units of the
   ! input array while [a] indicates the unscaled (e.g., mks) units that can be used with the reproducing sums
-  real, dimension(SZI_(G),SZJ_(G)) :: tmpForSumming ! The mass-weighted integral of the variable in a column [kg a]
-  real :: scalefac   ! An overall scaling factor for the cell mass and variable [a kg A-1 H-1 L-2 ~> kg m-3 or 1]
-  real :: temp_scale ! A temporary scaling factor [1] or [a A-1 ~> 1]
+  ! [A ~> a] and [B ~> b] are the same units unless tmp_scale and unscale are both present.
+  real, dimension(SZI_(G),SZJ_(G)) :: tmpForSumming ! The mass-weighted integral of the variable in a
+                     ! column [kg a] or [kg] or [B R Z L2 ~> kg b] or [R Z L2 ~> kg] if tmp_scale is present.
+  real :: scalefac   ! An overall scaling factor for the cell mass and variable [a kg A-1 R-1 Z-1 L-2 ~> 1]
+                     ! or [1] or [B A-1 ~> b a-1] if tmp_scale is present.
+  real :: temp_scale ! A temporary scaling factor [1] or [kg b R-1 Z-1 L-2 B-1 ~> 1] if tmp_scale is present.
   logical :: global_sum ! If true do the sum globally, but if false only do the sum on the current PE.
   integer :: i, j, k, is, ie, js, je, nz
   is = G%isc ; ie = G%iec ; js = G%jsc ; je = G%jec ; nz = GV%ke
 
-  temp_scale = 1.0 ; if (present(tmp_scale)) temp_scale = tmp_scale
-  scalefac = G%US%L_to_m**2*temp_scale
+  if (present(tmp_scale)) then
+    temp_scale = G%US%RZL2_to_kg * tmp_scale
+    scalefac = 1.0
+  else
+    temp_scale = 1.0
+    scalefac = G%US%RZL2_to_kg
+  endif
   if (present(unscale)) then ; scalefac = scalefac * unscale
   elseif (present(scale)) then ; scalefac = scalefac * scale ; endif
-  tmpForSumming(:,:) = 0.0
 
+  tmpForSumming(:,:) = 0.0
   if (present(var)) then
     do k=1,nz ; do j=js,je ; do i=is,ie
       tmpForSumming(i,j) = tmpForSumming(i,j) + var(i,j,k) * &
-                ((GV%H_to_kg_m2 * h(i,j,k)) * (scalefac*G%areaT(i,j) * G%mask2dT(i,j)))
+                ((GV%H_to_RZ * h(i,j,k)) * (scalefac*G%areaT(i,j) * G%mask2dT(i,j)))
     enddo ; enddo ; enddo
   else
     do k=1,nz ; do j=js,je ; do i=is,ie
       tmpForSumming(i,j) = tmpForSumming(i,j) + &
-                ((GV%H_to_kg_m2 * h(i,j,k)) * (scalefac*G%areaT(i,j) * G%mask2dT(i,j)))
+                ((GV%H_to_RZ * h(i,j,k)) * (scalefac*G%areaT(i,j) * G%mask2dT(i,j)))
     enddo ; enddo ; enddo
   endif
   global_sum = .true. ; if (present(on_PE_only)) global_sum = .not.on_PE_only
   if (global_sum) then
-    global_mass_integral = reproducing_sum(tmpForSumming)
+    global_mass_integral = reproducing_sum(tmpForSumming, unscale=temp_scale)
```

Comment:
See the similar comment above for function `global_area_integral`

Reply:
The reason here is the same, and the code is correct as written.